File: python/app/services/lists.py
"""lists collection: board-wide read granted to all three roles,
create/rename/delete/reorder restricted to owner only — enforced
server-side by Mudbase collection permissions (see plan/build-plan.md).
"""

import asyncio

from app.config import get_settings
from app.mudbase_client import create_data_sync, delete_data_sync, get_data_sync, list_data_sync, update_data_sync
from app.schemas.list_ import BoardList
from app.services.cards import list_cards_in_list
from app.services.cards import delete_card as delete_card_row

_LISTS_LIMIT = 100  # Mudbase's DataApi.list_data caps `limit` at 100 server-side.


async def list_board_lists(*, access_token: str) -> list[BoardList]:
    settings = get_settings()
    result = await asyncio.to_thread(
        list_data_sync,
        settings.lists_collection_id,
        filter_dict={"boardId": settings.board_id},
        sort="position",
        limit=_LISTS_LIMIT,
        access_token=access_token,
    )
    return [BoardList.model_validate(doc) for doc in result["data"]]


async def get_list(list_id: str, *, access_token: str) -> BoardList | None:
    settings = get_settings()
    doc = await asyncio.to_thread(get_data_sync, settings.lists_collection_id, list_id, access_token=access_token)
    return BoardList.model_validate(doc) if doc else None


async def next_list_position(*, access_token: str) -> float:
    lists = await list_board_lists(access_token=access_token)
    if not lists:
        return 0.0
    return max(board_list.position for board_list in lists) + 1.0


async def create_list(name: str, *, position: float, access_token: str) -> BoardList:
    settings = get_settings()
    doc = await asyncio.to_thread(
        create_data_sync,
        settings.lists_collection_id,
        {"boardId": settings.board_id, "name": name, "position": position},
        access_token=access_token,
    )
    return BoardList.model_validate(doc)


async def rename_list(board_list: BoardList, new_name: str, *, access_token: str) -> BoardList:
    settings = get_settings()
    doc = await asyncio.to_thread(
        update_data_sync,
        settings.lists_collection_id,
        board_list.id,
        {"name": new_name},
        access_token=access_token,
    )
    return BoardList.model_validate(doc)


async def reorder_list(board_list: BoardList, neighbor: BoardList, *, access_token: str) -> None:
    """Swaps this list's position with an adjacent one (left/right column
    reorder), mirroring ../web/src/hooks/useLists.ts::useReorderList."""
    settings = get_settings()
    await asyncio.to_thread(
        update_data_sync,
        settings.lists_collection_id,
        board_list.id,
        {"position": neighbor.position},
        access_token=access_token,
    )
    await asyncio.to_thread(
        update_data_sync,
        settings.lists_collection_id,
        neighbor.id,
        {"position": board_list.position},
        access_token=access_token,
    )


async def delete_list(board_list: BoardList, *, access_token: str) -> None:
    """Deletes a column and cascades to every card inside it — a list's
    cards have nowhere else to live once the column is gone, there's no
    "uncategorized" list to fall back to (mirrors
    ../web/src/hooks/useLists.ts::useDeleteList)."""
    settings = get_settings()
    cards_in_list = await list_cards_in_list(board_list.id, access_token=access_token)
    for card in cards_in_list:
        await delete_card_row(card, access_token=access_token)
    await asyncio.to_thread(delete_data_sync, settings.lists_collection_id, board_list.id, access_token=access_token)

File: go/internal/store/lists.go
// Package store implements every domain operation this app needs against the three Mudbase
// collections (lists, cards, activity), on top of internal/mbase's generic CRUD helpers. Every
// mutating method here also writes the corresponding `activity` row itself, so a handler can never
// forget to log an action - mirrors the reference web app's src/hooks/useLists.ts /
// src/hooks/useCards.ts, which pair every mutation with an activity write in the same hook.
package store

import (
	"context"
	"fmt"
	"sort"

	"github.com/mudbase/mudbase-showcase-kanban/go/internal/mbase"
	"github.com/mudbase/mudbase-showcase-kanban/go/internal/models"
)

// ListService implements every operation the app needs against the `lists` collection.
type ListService struct {
	client       *mbase.Client
	collectionID string
	activity     *ActivityService
}

// NewListService builds a ListService bound to the given lists collection ID.
func NewListService(client *mbase.Client, listsCollectionID string, activity *ActivityService) *ListService {
	return &ListService{client: client, collectionID: listsCollectionID, activity: activity}
}

// All returns every column on the shared board, ascending by position (column order).
func (s *ListService) All(ctx context.Context, token string) ([]models.List, error) {
	result, err := mbase.List[models.List](ctx, s.client, token, s.collectionID, mbase.ListParams{
		Filter: map[string]interface{}{"boardId": s.client.BoardID()},
		Sort:   "position",
		Limit:  200,
	})
	if err != nil {
		return nil, fmt.Errorf("store: listing lists: %w", err)
	}
	sort.Slice(result.Data, func(i, j int) bool { return result.Data[i].Position < result.Data[j].Position })
	return result.Data, nil
}

// ByID fetches a single list by its Mudbase document ID.
func (s *ListService) ByID(ctx context.Context, token, id string) (models.List, error) {
	list, err := mbase.Get[models.List](ctx, s.client, token, s.collectionID, id)
	if err != nil {
		return models.List{}, fmt.Errorf("store: fetching list %s: %w", id, err)
	}
	return list, nil
}

// nextPosition returns one past the highest position currently in use, 0 if there are none yet.
func nextPosition(lists []models.List) int {
	if len(lists) == 0 {
		return 0
	}
	max := lists[0].Position
	for _, l := range lists[1:] {
		if l.Position > max {
			max = l.Position
		}
	}
	return max + 1
}

// Create adds a new column at the end of the board and logs a created_list activity row.
// Owner-only - enforced by internal/server/middleware.go's requireRole before this is ever
// called, and independently by Mudbase's own collection permissions server-side.
func (s *ListService) Create(ctx context.Context, token, actorID, actorName, name string) (models.List, error) {
	existing, err := s.All(ctx, token)
	if err != nil {
		return models.List{}, err
	}

	created, err := mbase.Create[models.List](ctx, s.client, token, s.collectionID, map[string]interface{}{
		"boardId":  s.client.BoardID(),
		"name":     name,
		"position": nextPosition(existing),
	})
	if err != nil {
		return models.List{}, fmt.Errorf("store: creating list: %w", err)
	}

	if err := s.activity.Log(ctx, token, actorID, actorName, models.ActionCreatedList, "", "", name); err != nil {
		return models.List{}, err
	}
	return created, nil
}

// Rename updates a column's name and logs a renamed_list activity row. Owner-only.
func (s *ListService) Rename(ctx context.Context, token, actorID, actorName string, list models.List, newName string) (models.List, error) {
	updated, err := mbase.Update[models.List](ctx, s.client, token, s.collectionID, list.ID, map[string]interface{}{
		"name": newName,
	})
	if err != nil {
		return models.List{}, fmt.Errorf("store: renaming list %s: %w", list.ID, err)
	}

	if err := s.activity.Log(ctx, token, actorID, actorName, models.ActionRenamedList, "", list.Name, newName); err != nil {
		return models.List{}, err
	}
	return updated, nil
}

// Reorder swaps list's position with neighbor's - a left/right column move. Not itself logged to
// the activity feed, matching the reference web app's useReorderList. Owner-only.
func (s *ListService) Reorder(ctx context.Context, token string, list, neighbor models.List) error {
	if _, err := mbase.Update[models.List](ctx, s.client, token, s.collectionID, list.ID, map[string]interface{}{
		"position": neighbor.Position,
	}); err != nil {
		return fmt.Errorf("store: reordering list %s: %w", list.ID, err)
	}
	if _, err := mbase.Update[models.List](ctx, s.client, token, s.collectionID, neighbor.ID, map[string]interface{}{
		"position": list.Position,
	}); err != nil {
		return fmt.Errorf("store: reordering list %s: %w", neighbor.ID, err)
	}
	return nil
}

// Delete removes a column and every card inside it - a list's cards have nowhere else to live
// once the column is gone, there's no "uncategorized" list to fall back to (matches the reference
// web app's useDeleteList). Logs a deleted_list activity row. Owner-only.
func (s *ListService) Delete(ctx context.Context, token, actorID, actorName string, list models.List, cardsInList []models.Card, cards *CardService) error {
	for _, card := range cardsInList {
		if err := mbase.Delete(ctx, s.client, token, cards.collectionID, card.ID); err != nil {
			return fmt.Errorf("store: deleting card %s while deleting list %s: %w", card.ID, list.ID, err)
		}
	}
	if err := mbase.Delete(ctx, s.client, token, s.collectionID, list.ID); err != nil {
		return fmt.Errorf("store: deleting list %s: %w", list.ID, err)
	}

	return s.activity.Log(ctx, token, actorID, actorName, models.ActionDeletedList, "", list.Name, "")
}
